File: sdks/rs/src/internal/core/duplicate.rs
use serde_json::Value;

use super::duplicate_policy::DuplicatePolicy;

pub fn resolve_duplicate(
    existing: Value,
    incoming: Value,
    policy: DuplicatePolicy,
) -> Result<Value, String> {
    match policy {
        DuplicatePolicy::CanonicalWins | DuplicatePolicy::FirstWins => Ok(existing),
        DuplicatePolicy::UserWins | DuplicatePolicy::LastWins => Ok(incoming),
        DuplicatePolicy::KeepBoth => {
            let result = match existing {
                Value::Array(mut values) => {
                    values.push(incoming);
                    Value::Array(values)
                }
                other => Value::Array(vec![other, incoming]),
            };
            Ok(result)
        }
        DuplicatePolicy::ErrorOnDuplicate => {
            Err("duplicate field rejected by ErrorOnDuplicate policy".to_string())
        }
    }
}

File: sdks/py/src/loxa/core/duplicate.py
from __future__ import annotations

from typing import Any
from .duplicate_policy import CanonicalWins, ErrorOnDuplicate, FirstWins, KeepBoth, LastWins, UserWins


def apply_duplicate(target: dict[str, Any], key: str, value: Any, policy: str) -> None:
    if key not in target or policy in {UserWins, LastWins}:
        target[key] = value
        return
    if policy in {CanonicalWins, FirstWins}:
        return
    if policy == KeepBoth:
        old = target[key]
        target[key] = old + [value] if isinstance(old, list) else [old, value]
        return
    if policy == ErrorOnDuplicate:
        raise ValueError(f"duplicate field: {key}")
    target[key] = value
